src/schematools/events/partial.py

```diff
@@ -0,0 +1,549 @@
+        db_info = DbInfo.from_event_type(is_nm_table, event_type)
+
+        relation_handler = RelationHandler.fetch_handler(is_nm_table)(
+            source_id, event_data, event_type, self.tables, db_info, schema_info
+        )
+
+        relation_handler.set_query_info()
+        relation_handler.set_values()
+        relation_handler.execute(self.conn)
+
+    def process_row(self, source_id, event_meta, event_data):
+
+        event_type = event_meta["event_type"]
+        db_operation_name, needs_select, _, data_fetcher = EVENT_TYPE_MAPPINGS[NM_TABLE][
+            event_type
+        ]
+        row = {}
+        dataset_id = event_meta["catalog"]
+        table_id = event_meta["collection"]
+        if data_fetcher is not None:
+            row = data_fetcher(event_data)
+            for field_name in self.geo_fields[dataset_id][table_id]:
+                geo_value = row.get(field_name)
+                if geo_value is not None:
+                    row[field_name] = f"SRID={self.srid};{geo_value}"
+
+            # Only for ADD we need to generate the PK
+            if event_type == "ADD":
+                identifier = self.datasets[dataset_id].get_table_by_id(table_id).identifier
+                id_value = ".".join(str(row[fn]) for fn in identifier)
+                row["id"] = id_value
+            row["source_id"] = source_id
+
+        table = self.tables[dataset_id][table_id]
+        db_operation = getattr(table, db_operation_name)()
+        if needs_select:
+            db_operation = db_operation.where(table.c.source_id == source_id)
+        self.conn.execute(db_operation, row)
+
+    def process_event(self, source_id, event_meta, event_data, is_relation=False):
+        """ Do inserts/updates/deletes """
+
+        if is_relation:
+            try:
+                self.process_relation(source_id, event_meta, event_data)
+            except UnknownRelationException as e:
+                logger.warning("Unknown Relation: %s", e)
+        else:
+            self.process_row(source_id, event_meta, event_data)
+
+    def load_events_from_file(self, events_path):
+        """ Helper method, primarily used for testing """
+
+        with open(events_path) as ef:
+            for line in ef:
+                if line.strip():
+                    source_id, event_meta_str, data_str = line.split("|", maxsplit=2)
+                    event_meta = json.loads(event_meta_str)
+                    is_relation = event_meta["catalog"] == "rel"
+                    event_data = json.loads(data_str)
+                    self.process_event(
+                        source_id,
+                        event_meta,
+                        event_data,
+                        is_relation,
+                    )
+
+
+def tables_factory(
+    dataset: DatasetSchema,
+    metadata: Optional[MetaData] = None,
+) -> Dict[str, Table]:
+    """Generate the SQLAlchemy Table objects to work with the JSON Schema
```

Comment:
Dit lijkt me super generiek. Waarom niet in een aparte file? of bij andere meer DB-low-level code i.p.v. ten midden van de events.